File: L3/App_3_2/App_3_2.py
# 
# PCLP1 | L3
# Calcularea sumei S si determinarea err.
# 
# File: App_3_2.py
# 

from math import pow, sin

def fact(n):
    if n == 0:
        return 1

    return n * fact(n - 1)

def s(x, N):
    if N <= 0:
        raise Exception("N trebuie sa fie pozitiv")

    suma = 0

    for i in range(N - 1):
        suma = suma + (pow(-1, i) / fact(2 * i + 1)) * pow(x, 2 * i + 1)

    return suma 

def main():
    N = int(input("  N = "))
    x = float(input("  x = "))

    try:
        suma = s(x, N)
        print(f"  s = {suma}")

        err = abs((suma - sin(x)) / sin(x))
        print(f"  err = {err}")
    except Exception as e:
        print(f"  Error: {e}")

if __name__ == "__main__":
    main()

File: L3/App_3_2/App_3_2.cpp
/**
 * @brief PCLP1 | L3
 *      Calcularea sumei S si determinarea err.
 *
 * @file App_3_2.cpp
 */

#include <iostream>
#include <cmath>

double fact(unsigned int n)
{
    if (n == 0)
        return 1;

    return n * fact(n - 1);
}

double s(double x, int N)
{
    if (N <= 0)
        throw std::runtime_error("N trebuie sa fie pozitiv!\n");

    double suma = 0;

    for (unsigned int i = 0; i < N; ++i)
        suma += (std::pow(-1, i) / fact(2 * i + 1)) * std::pow(x, 2 * i + 1);

    return suma;
}

int main()
{
    int N = 0;
    double x = 0;

    std::cout << "  N = "; std::cin >> N;
    std::cout << "  x = "; std::cin >> x;

    try
    {
        double suma = s(x, N);
        
        std::cout << "  s = " << suma << '\n';
        std::cout << "  err = " << std::abs((suma - std::sin(x)) / std::sin(x)) << '\n';
    }
    catch(const std::exception &e)
    {
        std::cerr << "  Error: " << e.what() << "\n";
    }

    return 0;
}